### PR TITLE
Source Codecs | Avro Codec follow-on PR

### LINKED ISSUE #1532: Support generic parsers/codecs
**Is your feature request related to a problem? Please describe.**

Both the [S3 Source](https://github.com/opensearch-project/data-prepper/issues/251) and HTTP Source use similar concepts of codecs for parsing input data. The S3 Source currently makes these codecs available as plugins. So they can be extended for the S3 source. But, if another source wanted to use these plugins it would be unable to.

**Describe the solution you'd like**

Create a core concept in Data Prepper of source-based codecs or parsers. These should be generic enough to take any Java `InputStream` and produce events from them.

I propose that we based this concept on the S3 codec. It has a few advantages:

1. It uses an InputStream. This is advantageous for large inputs2. 
2. It has a `Consumer` for each event. This also allows the source using the codec to receive `Event` objects and decide independently of the Codec of the best way to handle these.
3. It is not connected to HTTP in anyway.

**Describe alternatives you've considered (Optional)**

Data Prepper can have a similar concept for output codecs/parsers. However, I see no reason to force these to be the same concept. (Implementors may choose to pair them together to avoid code duplication).

**Additional context**

S3 Codec interface:

https://github.com/opensearch-project/data-prepper/blob/37c8b09ae8b7b8ce233564f5da7b09d1d41953f2/data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/Codec.java#L19-L27

HTTP Codec interface:

https://github.com/opensearch-project/data-prepper/blob/37c8b09ae8b7b8ce233564f5da7b09d1d41953f2/data-prepper-plugins/http-source/src/main/java/com/amazon/dataprepper/plugins/source/loghttp/codec/Codec.java#L16-L23

### PR DESCRIPTION
**Pull request by umayr-codes**
### Description
Follow on PR on Avro codecs with support for nested and other non-primitive data types.
 
### Issues Resolved
Resolves #1532 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/avro-codecs/src/test/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodecTest.java [hunk 1]
```diff
@@ -166,18 +168,40 @@ private static Object decodeOutputIfEncoded(Map encodedOutput, Object key){
...
         List<GenericRecord> recordList=generateRecords(parseSchema(),numberOfRecords);
         GenericRecord record=recordList.get(index);
-        Schema schema=parseSchema();
+        Schema schema = record.getSchema();
+        final Map<String, Object> eventData = convertRecordToMap(record, schema);
+        final Event event = JacksonLog.builder().withData(eventData).build();
+        return event;
+    }
+
+    private static Map<String, Object> convertRecordToMap(GenericRecord record, Schema schema) throws Exception {
```
Comment (dlvenable) [inline data-prepper-plugins/avro-codecs/src/test/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodecTest.java:180]: You are duplicating the logic from `AvroInputCodec` in the test. This is not a valid test scenario. You should start with known data coming in and verify it against expected data going out.

Please remove this entire function - `convertRecordToMap`.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/avro-codecs/src/test/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodecTest.java (hunk 1 above)
Comment (umayr-codes) [reply data-prepper-plugins/avro-codecs/src/test/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodecTest.java:180]: Okay...

Comment (umayr-codes) [reply data-prepper-plugins/avro-codecs/src/test/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodecTest.java]: It is done @dlvenable !

**Comment by dlvenable** [comment]
@umayr-codes , Did you mean to include work for the output codecs in this PR?

**Comment by umayr-codes** [comment]
> @umayr-codes , Did you mean to include work for the output codecs in this PR?

No. That commit has been rolled back.

**Review by dlvenable — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]

**Pull request merged**